csharp/src/Ice/Communicator-ObjectAdapterFactory.cs

```diff
@@ -17,30 +17,17 @@ public async Task ShutdownAsync()
         {
             lock (_mutex)
             {
-                // _shutdownSemaphore != null means "shutdown in progress". Once shutdown is in progress, other methods
-                // cannot modify _adapters.
-                _shutdownSemaphore ??= new SemaphoreSlim(1, 1);
+                _shutdown = true;
                 _waitForShutdownCompletionSource ??= new TaskCompletionSource<object?>();
             }
 
-            // The first call that acquires the semaphore is the one that calls DisposeAsync on the adapters.
-            await _shutdownSemaphore.WaitAsync().ConfigureAwait(false);
-
             try
             {
-                // _adapters can only be changed by the first call that acquires the semaphore.
-                await Task.WhenAll(_adapters.Select(adapter => adapter.DisposeAsync().AsTask())).ConfigureAwait(false);
+                // _adapters can only be updated when _shutdown is false
```

Comment:
// Don't need to lock _mutex, _adapters can only ....

Reply:
Fixed.